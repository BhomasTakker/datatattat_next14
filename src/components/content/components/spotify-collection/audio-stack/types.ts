import { ContainerHeightOptions } from "@/components/page/components/stack/types";
import { SearchType } from "@/types/api/spotify";
import { CollectionItem } from "@/types/data-structures/collection/item/item";

export type SpotifyCollectionItem = CollectionItem & {
	id: string;
	media: {
		type: SearchType;
	};
};

export type AudioStackClientProps = {
	items: SpotifyCollectionItem[];
	height: ContainerHeightOptions;
};

export type SpotifyArticleProps = {
	item: SpotifyCollectionItem;
	onClick: (index: number) => void;
	index: number;
};
