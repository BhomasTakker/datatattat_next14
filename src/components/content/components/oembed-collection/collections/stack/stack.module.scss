.root {
	.list {
		display: flex;
		flex-direction: column;
		align-items: center;
		width: 100%;
		li {
			// sm, md, etc
			width: 100%;
		}
	}
}
