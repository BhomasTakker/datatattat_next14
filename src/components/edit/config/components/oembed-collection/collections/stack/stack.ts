import { EditInputs } from "@/components/edit/inputs/inputs";
import { InputListProps } from "@/types/edit/inputs/inputs";

export const OEMBED_STACK_CONFIG: InputListProps = {
	id: "stack",
	type: EditInputs.inputList,
	label: "Stack",
	inputs: [
		{
			id: "oembedComponentTitle",
			type: EditInputs.title,
			title: "Oembed Stack Component",
		},
	],
};
