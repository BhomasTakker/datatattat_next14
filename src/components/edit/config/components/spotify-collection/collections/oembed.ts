import { EditInputs } from "@/components/edit/inputs/inputs";
import { InputListProps } from "@/types/edit/inputs/inputs";
import { getWithConfig, QueryOptions } from "../../../query/_with-config";

export const SPOTIFY_OEMBED_CONFIG: InputListProps = {
	id: "spotifyOembed",
	type: EditInputs.inputList,
	label: "Spotify Oembed",
	inputs: [
		{
			id: "oembedComponentTitle",
			type: EditInputs.title,
			title: "Spotify Oembed Component",
		},
		getWithConfig({
			options: [
				// we need to  add this after we determine variant
				QueryOptions.SPOTIFY_OEMBED,
			],
			defaultSelection: QueryOptions.SPOTIFY_OEMBED,
		}),
	],
};
