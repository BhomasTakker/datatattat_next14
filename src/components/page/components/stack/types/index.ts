import { PageStackCollectionVariants } from "@/types/components/page/stack";

export type VerticalStackProps = {
	variant: PageStackCollectionVariants.Vertical;
};

export const ContainerWidth = {
	XXS: "180",
	XS: "240",
	SM: "320",
	MD: "640",
	LG: "768",
	XL: "1024",
	XXL: "1280",
} as const;

export const ContainerWidthOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerWidth = (typeof ContainerWidth)[keyof typeof ContainerWidth];

export type ContainerWidthOptions = keyof typeof ContainerWidth;

export const ContainerHeight = {
	XXS: "240",
	XS: "360",
	SM: "480",
	MD: "600",
	LG: "720",
	XL: "840",
	XXL: "960",
} as const;

export const ContainerHeightOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerHeight = (typeof ContainerHeight)[keyof typeof ContainerHeight];

type ContainerHeightOptions = keyof typeof ContainerHeight;
export const containerWidthOptions: ContainerWidthOptions[] = Object.keys(
	ContainerWidth
) as ContainerWidthOptions[];

// Row configuration for row-stack layout
export type Row = {
	columns: number;
	evenColumns: boolean;
	// you wouldn't max/min height - you would set a height
	maxHeight: ContainerHeightOptions;
	minWidth: ContainerWidthOptions;
	// these are all actually strings!
	// Unless we can/want to convert these after http requests use strings
	index?: string; // optional for default row - specifies which row this config applies to
};

export type Column = {
	minWidth: ContainerWidthOptions;
	maxHeight: ContainerHeightOptions;
	evenColumns: boolean;
};

export type RowStackProps = {
	variant: PageStackCollectionVariants.RowStack;
	defaultRow: Row; // Default configuration used for any row not specifically configured
	rows?: Row[]; // Optional array of row-specific configurations - index property determines which row each config applies to
};

export type PageStackProps = {
	variant: PageStackCollectionVariants;
} & (VerticalStackProps | RowStackProps);
