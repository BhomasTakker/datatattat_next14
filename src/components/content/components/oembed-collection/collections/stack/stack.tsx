import { OEmbed } from "@/types/data-structures/oembed";
import { ClientOembed } from "../../../content-oembed/client-component";
import styles from "./stack.module.scss";

export type OembedStackProps = {};

const renderOembed = (item: OEmbed) => {
	const { src, html } = item || {};

	return (
		<li key={item.id} data-testid="content-oembed" className={styles.root}>
			{html && <ClientOembed html={html} />}
		</li>
	);
};

const renderMethod = (collection: OEmbed[] = [], props: OembedStackProps) => {
	return (
		<ul className={`${styles.list}`}>
			{collection.map((item) => renderOembed(item))}
		</ul>
	);
};

const oembedStack = {
	styles,
	renderMethod,
};

export default oembedStack;
