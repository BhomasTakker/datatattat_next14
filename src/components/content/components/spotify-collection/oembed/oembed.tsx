import { OEmbed } from "@/types/data-structures/oembed";
import styles from "./oembed.module.scss";
import { SpotifyOembedComponent } from "./spotify-oembed-component";

export type OembedStackProps = {};

type SpotifyOembed = {
	iframe_url: string;
} & OEmbed;

const renderMethod = (items: SpotifyOembed[] = [], props: OembedStackProps) => {
	const [item] = items;

	return <SpotifyOembedComponent item={item} />;
};

const spotifyOembed = {
	styles,
	renderMethod,
};

export default spotifyOembed;
