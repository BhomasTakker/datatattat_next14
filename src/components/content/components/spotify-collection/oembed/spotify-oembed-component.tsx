import { ClientOembed } from "../../oembed-collection/content-oembed/client-component";
import { OEmbed } from "@/types/data-structures/oembed";
import {
	ContainerHeightOptions,
	getContainerHeight,
} from "@/components/page/components/stack/types";

const getOembedType = (item: SpotifyOembed) => {
	const { iframe_url = "" } = item;
	let oembedType;

	switch (true) {
		case iframe_url.includes("embed/album"):
			oembedType = "album";
			break;
		case iframe_url.includes("embed/artist"):
			oembedType = "artist";
			break;
		case iframe_url.includes("embed/track"):
			oembedType = "track";
			break;
		case iframe_url.includes("embed/show"):
			oembedType = "show";
			break;
		case iframe_url.includes("embed/playlist"):
			oembedType = "playlist";
			break;
		case iframe_url.includes("embed/episode"):
			oembedType = "episode";
			break;
	}

	return oembedType;
};

type SpotifyOembed = {
	iframe_url: string;
} & OEmbed;

type SpotifyOembedComponentProps = {
	item: SpotifyOembed;
	height: ContainerHeightOptions;
};

export const SpotifyOembedComponent = ({
	item,
	height = ContainerHeightOptions.MD,
}: SpotifyOembedComponentProps) => {
	const { html = "" } = item;
	let oembedType = getOembedType(item);

	const updateSize = ["album", "artist", "playlist"].includes(oembedType || "");

	const containerHeight = getContainerHeight(height);

	const newHtml = html.replace(/height="\d+"/, `height="${containerHeight}"`);
	const useHtml = updateSize ? newHtml : html;

	return (
		<div>
			<ClientOembed html={useHtml} />
		</div>
	);
};
