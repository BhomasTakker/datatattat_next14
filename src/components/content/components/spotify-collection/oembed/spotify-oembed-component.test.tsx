import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { SpotifyOembedComponent } from "./spotify-oembed-component";
import { ContainerHeightOptions } from "../../../../../components/page/components/stack/types";
import { OEmbed } from "@/types/data-structures/oembed";

// Mock the ClientOembed component
jest.mock("../../oembed-collection/content-oembed/client-component", () => ({
	ClientOembed: ({ html }: { html: string }) => (
		<div
			data-testid="client-oembed"
			dangerouslySetInnerHTML={{ __html: html }}
		/>
	),
}));

// Mock the getContainerHeight function
jest.mock("../../../../../components/page/components/stack/types", () => ({
	...jest.requireActual(
		"../../../../../components/page/components/stack/types"
	),
	getContainerHeight: jest.fn((height: string) => {
		const heights: Record<string, string> = {
			XXS: "120",
			XS: "200",
			SM: "280",
			MD: "360",
			LG: "480",
			XL: "600",
			XXL: "720",
		};
		return heights[height] || "360";
	}),
}));

type SpotifyOembed = {
	iframe_url: string;
} & OEmbed;

const createMockSpotifyOembed = (
	iframe_url: string,
	html: string = '<iframe src="test" height="152"></iframe>'
): SpotifyOembed => ({
	id: "test-id",
	title: "Test Spotify Content",
	type: "rich",
	version: "1.0",
	provider_name: "Spotify",
	provider_url: "https://spotify.com",
	html,
	iframe_url,
	width: 300,
	height: 152,
});

describe("SpotifyOembedComponent", () => {
	describe("renders correctly with different Spotify content types", () => {
		it("renders album content", () => {
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/album/123"
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.MD}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			expect(clientOembed).toBeInTheDocument();
		});

		it("renders artist content", () => {
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/artist/456"
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.MD}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			expect(clientOembed).toBeInTheDocument();
		});

		it("renders track content", () => {
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/track/789"
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.MD}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			expect(clientOembed).toBeInTheDocument();
		});

		it("renders show content", () => {
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/show/abc"
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.MD}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			expect(clientOembed).toBeInTheDocument();
		});

		it("renders playlist content", () => {
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/playlist/def"
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.MD}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			expect(clientOembed).toBeInTheDocument();
		});

		it("renders episode content", () => {
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/episode/ghi"
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.MD}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			expect(clientOembed).toBeInTheDocument();
		});
	});

	describe("height adjustment functionality", () => {
		it("updates height for album content type", () => {
			const originalHtml = '<iframe src="test" height="152"></iframe>';
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/album/123",
				originalHtml
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.LG}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			expect(clientOembed.innerHTML).toContain('height="480"');
		});

		it("updates height for artist content type", () => {
			const originalHtml = '<iframe src="test" height="152"></iframe>';
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/artist/456",
				originalHtml
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.XL}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			expect(clientOembed.innerHTML).toContain('height="600"');
		});

		it("updates height for playlist content type", () => {
			const originalHtml = '<iframe src="test" height="152"></iframe>';
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/playlist/def",
				originalHtml
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.SM}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			expect(clientOembed.innerHTML).toContain('height="280"');
		});

		it("does not update height for track content type", () => {
			const originalHtml = '<iframe src="test" height="152"></iframe>';
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/track/789",
				originalHtml
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.LG}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			expect(clientOembed.innerHTML).toContain('height="152"');
			expect(clientOembed.innerHTML).not.toContain('height="480"');
		});

		it("does not update height for show content type", () => {
			const originalHtml = '<iframe src="test" height="152"></iframe>';
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/show/abc",
				originalHtml
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.LG}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			expect(clientOembed.innerHTML).toContain('height="152"');
			expect(clientOembed.innerHTML).not.toContain('height="480"');
		});

		it("does not update height for episode content type", () => {
			const originalHtml = '<iframe src="test" height="152"></iframe>';
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/episode/ghi",
				originalHtml
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.LG}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			expect(clientOembed.innerHTML).toContain('height="152"');
			expect(clientOembed.innerHTML).not.toContain('height="480"');
		});
	});

	describe("default props", () => {
		it("uses default height when not provided", () => {
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/album/123"
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.MD}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			expect(clientOembed).toBeInTheDocument();
		});
	});

	describe("edge cases", () => {
		it("handles empty html gracefully", () => {
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/album/123",
				""
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.MD}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			expect(clientOembed).toBeInTheDocument();
			expect(clientOembed.innerHTML).toBe("");
		});

		it("handles iframe_url without known embed type", () => {
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/unknown/123",
				'<iframe src="test" height="152"></iframe>'
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.LG}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			// Should not update height since it's not a known type
			expect(clientOembed.innerHTML).toContain('height="152"');
			expect(clientOembed.innerHTML).not.toContain('height="480"');
		});

		it("handles html without height attribute", () => {
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/album/123",
				'<iframe src="test"></iframe>'
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.LG}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			expect(clientOembed).toBeInTheDocument();
		});

		it("handles html with multiple height attributes correctly", () => {
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/album/123",
				'<iframe src="test" height="152" style="height: 200px;"></iframe>'
			);

			render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.LG}
				/>
			);

			const clientOembed = screen.getByTestId("client-oembed");
			// Should only replace the height attribute, not the style
			expect(clientOembed.innerHTML).toContain('height="480"');
			expect(clientOembed.innerHTML).toContain('style="height: 200px;"');
		});
	});

	describe("component structure", () => {
		it("wraps ClientOembed in a div", () => {
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/track/789"
			);

			const { container } = render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.MD}
				/>
			);

			const wrapper = container.firstChild;
			expect(wrapper).toBeInstanceOf(HTMLDivElement);
			expect(wrapper).toContainElement(screen.getByTestId("client-oembed"));
		});

		it("matches snapshot for album content", () => {
			const mockItem = createMockSpotifyOembed(
				"https://open.spotify.com/embed/album/123",
				'<iframe src="https://open.spotify.com/embed/album/123" height="152"></iframe>'
			);

			const { container } = render(
				<SpotifyOembedComponent
					item={mockItem}
					height={ContainerHeightOptions.MD}
				/>
			);

			expect(container).toMatchSnapshot();
		});
	});
});
