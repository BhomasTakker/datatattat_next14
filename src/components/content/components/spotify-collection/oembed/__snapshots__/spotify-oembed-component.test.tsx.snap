// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SpotifyOembedComponent component structure matches snapshot for album content 1`] = `
<div>
  <div>
    <div
      data-testid="client-oembed"
    >
      <iframe
        height="360"
        src="https://open.spotify.com/embed/album/123"
      />
    </div>
  </div>
</div>
`;
